src/Microsoft.Azure.ServiceBus.KeyVault/KeyVaultPlugin.cs

```diff
@@ -0,0 +1,166 @@
+            {
+                throw new ArgumentNullException(nameof(encryptionSecretName));
+            }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            this.secretName = encryptionSecretName;
+            this.keyVaultEndpoint = options.Endpoint;
+            this.secretManager = new KeyVaultSecretManager(options.Endpoint, options.ClientId, options.ClientSecret);
+            this.initializationVector = KeyVaultPlugin.GenerateInitializationVector();
+            this.base64InitializationVector = Convert.ToBase64String(this.initializationVector);
+        }
+
+        /// <summary>
+        /// The action performed before sending a message to Service Bus. This method will load the KeyVault key and encrypt messages.
+        /// </summary>
+        /// <param name="message">The <see cref="Message"/> to be encrypted.</param>
+        /// <returns>The encrypted <see cref="Message"/>.</returns>
+        public override async Task<Message> BeforeMessageSend(Message message)
+        {
+            try
+            {
+                if (message.UserProperties.ContainsKey(KeyVaultMessageHeaders.InitializationVectorPropertyName) || message.UserProperties.ContainsKey(KeyVaultMessageHeaders.KeyNamePropertyName))
+                {
+                    return message;
+                }
+
+                var secret = await secretManager.GetHashedSecret(secretName);
+
+                message.UserProperties.Add(KeyVaultMessageHeaders.InitializationVectorPropertyName, base64InitializationVector);
+                message.UserProperties.Add(KeyVaultMessageHeaders.KeyNamePropertyName, secretName);
+
+                message.Body = await KeyVaultPlugin.Encrypt(message.Body, secret, this.initializationVector);
+                return message;
+            }
+            catch (Exception ex)
+            {
+                throw new KeyVaultPluginException(Resources.BeforeMessageSendException, ex);
+            }
+        }
+
+        /// <summary>
+        /// The action performed after receiving a message from Service Bus. This method will load the KeyVault key and decrypt messages.
+        /// </summary>
+        /// <param name="message">The <see cref="Message"/> to be decrypted.</param>
+        /// <returns>The decrypted <see cref="Message"/>.</returns>
+        public override async Task<Message> AfterMessageReceive(Message message)
+        {
+            try
+            {
+                if (!message.UserProperties.ContainsKey(KeyVaultMessageHeaders.InitializationVectorPropertyName) || !message.UserProperties.ContainsKey(KeyVaultMessageHeaders.KeyNamePropertyName))
+                {
+                    return message;
+                }
+
+                var iVString = message.UserProperties[KeyVaultMessageHeaders.InitializationVectorPropertyName] as string;
+                var iV = Convert.FromBase64String(iVString);
```

Comment:
>FromBase64String [](start = 33, length = 16)

why this conversion? where are you converting to string? Is it the `as string` conversion in the above line?
If so, can you directly use the byte[] array from the message?